cmake/CopyBuildProducts.cmake

```diff
@@ -0,0 +1,98 @@
+      list(APPEND "COPY_BUILD_PRODUCTS_${current_dependency}_FILES" "${arg}")
+    elseif("${mode}" STREQUAL "TARGET_LOCATION")
+      if(DEFINED "COPY_BUILD_PRODUCTS_${current_dependency}_TARGET_LOCATION")
+        message(FATAL_ERROR "Target location for dependency ${current_dependency} already set!")
+      endif()
+      set("COPY_BUILD_PRODUCTS_${current_dependency}_TARGET_LOCATION" "${arg}")
+    elseif("${mode}" STREQUAL "CONDITION")
+      if(DEFINED "COPY_BUILD_PRODUCTS_${current_dependency}_CONDITION")
+       set("COPY_BUILD_PRODUCTS_${current_dependency}_CONDITION" "${COPY_BUILD_PRODUCTS_${current_dependency}_CONDITION} ${arg}")
+      else()
+        set("COPY_BUILD_PRODUCTS_${current_dependency}_CONDITION" "${arg}")
+      endif()
+    endif()
+  endforeach()
+
+  set(COPY_BUILD_PRODUCTS_COPIED_RESOURCES)
+  foreach(dependency_idx RANGE 0 ${current_dependency})
+    list(GET COPY_BUILD_PRODUCTS_DEPENDENCIES ${dependency_idx} dependency)
+    get_target_property(type ${dependency} TYPE)
+    get_target_property(resx_tmp ${dependency} RESOURCE)
+    get_target_property(bindir ${dependency} BINARY_DIR)
+
+    set(resx "")
+    if(resx_tmp)
+      foreach(res IN LISTS resx_tmp)
+        if(IS_ABSOLUTE "${res}")
+          file(RELATIVE_PATH relpath "${bindir}" "${res}")
+          list(APPEND resx "${relpath}")
+        else()
+          list(APPEND resx "${res}")
+        endif()
+      endforeach()
+    endif()
+
+    set(cond_succeeded NO)
+    if(NOT DEFINED "COPY_BUILD_PRODUCTS_${dependency_idx}_CONDITION")
+      set(cond_succeeded YES)
+    else()
+      cmake_language(EVAL CODE "if(COPY_BUILD_PRODUCTS_${dependency_idx}_CONDITION)\nset(cond_succeeded YES)\nendif()")
+    endif()
+
+    if("${type}" STREQUAL "INTERFACE_LIBRARY")
+      set(tgt_dir "$<TARGET_PROPERTY:${dependency},BINARY_DIR>")
+      set(tgt_outputs "")
+    else()
+      set(tgt_dir "$<TARGET_FILE_DIR:${dependency}>")
+      set(tgt_outputs "$<TARGET_FILE_NAME:${dependency}>")
+    endif()
+
+    list(APPEND tgt_outputs ${resx})
+
+    if(DEFINED COPY_BUILD_PRODUCTS_${dependency_idx}_FILES)
+      set(tgt_outputs "${COPY_BUILD_PRODUCTS_${dependency_idx}_FILES}")
+    endif()
+
+    list(REMOVE_DUPLICATES tgt_outputs)
+
+    if(cond_succeeded)
+      foreach(file IN LISTS tgt_outputs)
+        list(APPEND COPY_BUILD_PRODUCTS_COPIED_RESOURCES "${COPY_BUILD_PRODUCTS_${dependency_idx}_TARGET_LOCATION}/${file}")
+        list(APPEND copy_snippet
+"# ${COPY_BUILD_PRODUCTS_${dependency_idx}_CONDITION} (${dependency} -> ${target})
+execute_process(COMMAND \${CMAKE_COMMAND} -E copy_if_different
+  ${tgt_dir}/${file}
+  ${COPY_BUILD_PRODUCTS_${dependency_idx}_TARGET_LOCATION}/${file}
+)")
+      endforeach()
+    endif()
+  endforeach()
+
+  list(JOIN copy_snippet "\n" copy_snippet)
+
+  get_property(multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+  file(GENERATE
+    OUTPUT copy_build_products$<$<BOOL:${multi_config}>:_$<CONFIG>>.cmake
```

Comment:
This `$<$<BOOL:${multi_config}>:_$<CONFIG>>` nonsense evaluates to `_$<CONFIG>` if the build-system generator supports multiple configurations, avoiding a CMake error where we try to overwrite the same file multiple times with different content. This means that we end up generating multiple files, one of which is selected based on the configuration. For single-configuration generators, this isn't necessary, so the `_$<CONFIG>` suffix is dropped entirely.